File: unit/test_crypto.py
"""Test crypto functions."""

import binascii
import os
import sys
import unittest

sys.path.append('..')  # noqa

# pylint: disable=C0413
from lib.crypto import (CryptHelper, ECDSAPrivateKey, ECDSAPublicKey, Hasher)


class TestCrypto(unittest.TestCase):
    """Test crypto stuff."""

    def verify_roundtrip(self, data):
        """Verify the crypto round-trip using specified data."""
        k, d = CryptHelper.encrypt_json(data)
        value = d.decrypt_json(k)
        self.assertEqual(value, data,
                         "decrypted data should match original")
        return

    def remove_if_exists(self, fn):
        """Delete file if it exists."""
        if os.path.exists(fn):
            os.remove(fn)
        return

    def test_roundtrip(self):
        """Test that AES can round-trip ok."""
        self.verify_roundtrip("The cat sat on the mat")
        self.verify_roundtrip({'key': 'value'})
        self.verify_roundtrip(["hello", "one", "two", "three"])
        return

    def test_ecdsa_sig(self):
        """Test signing of messages."""
        # Generate sig.
        msg = "The cat sat on the mat"
        h = Hasher(msg)
        key = ECDSAPrivateKey.generate()
        sig = h.sign(key)

        # Verify.
        pubkey = key.publickey
        h.verify_signature(sig, pubkey)
        return

    def test_ecdsa_sig_save_load(self):
        """Test signing of messages including saving to a file."""
        # Generate sig.
        msg = "The cat sat on the mat"
        h = Hasher(msg)
        key = ECDSAPrivateKey.generate()
        sig = h.sign(key)

        # Save to a file.
        password = "passphrase"
        fnbase = "keys"
        self.remove_if_exists(fnbase + "_private.pem")
        self.remove_if_exists(fnbase + "_public.pem")
        fnpriv, _ = key.write_key_pair(fnbase, password)

        # Load from file and verify.
        key2 = ECDSAPrivateKey.from_file(fnpriv, password)
        pubkey = key2.publickey
        h.verify_signature(sig, pubkey)
        return

    def test_ecdsa_serialise(self):
        """Test signing of messages including saving to a file."""
        # Generate sig.
        msg = "The cat sat on the mat"
        h = Hasher(msg)
        key = ECDSAPrivateKey.generate()
        sig = h.sign(key)

        # Convert to string.
        pk_str = key.to_string()

        # Convert back to pk and verify.
        key2 = ECDSAPrivateKey.from_string(pk_str)
        pubkey = key2.publickey
        h.verify_signature(sig, pubkey)
        return

    def test_ecdsa_copy_pubkey(self):
        """Test signing of messages including copying the pubkey."""
        # Generate sig.
        msg = "The cat sat on the mat"
        h = Hasher(msg)
        key = ECDSAPrivateKey.generate()
        sig = h.sign(key)

        pubkey1 = key.publickey.as_hash()
        pubkey2 = ECDSAPublicKey(pubkey1)
        h.verify_signature(sig, pubkey2)
        return

    def test_hasher(self):
        """Test the hasher."""
        h = Hasher()
        h.update("hello")
        digest = h.get_hash().digest()
        hexdigest = h.get_hash().to_string()
        self.assertEqual(binascii.hexlify(digest).decode('ascii'), hexdigest,
                         "hexdigest matches digest")
        return

File: rs/tests/test_crypto.rs
/// Crypto tests — direct port of unit/test_crypto.py.
use stevecoin::crypto::{CryptHelper, ECDSAPrivateKey, ECDSAPublicKey, Hasher};

fn remove_if_exists(path: &str) {
    let _ = std::fs::remove_file(path);
}

#[test]
fn test_roundtrip_string() {
    let data = serde_json::json!("The cat sat on the mat");
    let (k, enc) = CryptHelper::encrypt_json(&data).unwrap();
    let decrypted = enc.decrypt_json(&k).unwrap();
    assert_eq!(data, decrypted, "decrypted data should match original");
}

#[test]
fn test_roundtrip_dict() {
    let data = serde_json::json!({"key": "value"});
    let (k, enc) = CryptHelper::encrypt_json(&data).unwrap();
    let decrypted = enc.decrypt_json(&k).unwrap();
    assert_eq!(data, decrypted, "decrypted data should match original");
}

#[test]
fn test_roundtrip_list() {
    let data = serde_json::json!(["hello", "one", "two", "three"]);
    let (k, enc) = CryptHelper::encrypt_json(&data).unwrap();
    let decrypted = enc.decrypt_json(&k).unwrap();
    assert_eq!(data, decrypted, "decrypted data should match original");
}

#[test]
fn test_ecdsa_sig() {
    let msg = "The cat sat on the mat";
    let key = ECDSAPrivateKey::generate();
    let sig = Hasher::new_with_message(msg.as_bytes()).sign(&key).unwrap();

    let pubkey = key.publickey();
    Hasher::new_with_message(msg.as_bytes())
        .verify_signature(&sig, &pubkey)
        .unwrap();
}

#[test]
fn test_ecdsa_sig_save_load() {
    let msg = "The cat sat on the mat";
    let key = ECDSAPrivateKey::generate();
    let sig = Hasher::new_with_message(msg.as_bytes()).sign(&key).unwrap();

    let password = "passphrase";
    let fnbase = "/tmp/stevecoin_test_keys";
    remove_if_exists(&format!("{fnbase}_private.pem"));
    remove_if_exists(&format!("{fnbase}_public.pem"));

    let (fnpriv, _) = key.write_key_pair(fnbase, password).unwrap();

    let key2 = ECDSAPrivateKey::from_file(&fnpriv, password).unwrap();
    let pubkey = key2.publickey();
    Hasher::new_with_message(msg.as_bytes())
        .verify_signature(&sig, &pubkey)
        .unwrap();

    // Cleanup
    remove_if_exists(&format!("{fnbase}_private.pem"));
    remove_if_exists(&format!("{fnbase}_public.pem"));
}

#[test]
fn test_ecdsa_serialise() {
    let msg = "The cat sat on the mat";
    let key = ECDSAPrivateKey::generate();
    let sig = Hasher::new_with_message(msg.as_bytes()).sign(&key).unwrap();

    let pk_str = key.to_hex_string();

    let key2 = ECDSAPrivateKey::from_string(&pk_str).unwrap();
    let pubkey = key2.publickey();
    Hasher::new_with_message(msg.as_bytes())
        .verify_signature(&sig, &pubkey)
        .unwrap();
}

#[test]
fn test_ecdsa_copy_pubkey() {
    let msg = "The cat sat on the mat";
    let key = ECDSAPrivateKey::generate();
    let sig = Hasher::new_with_message(msg.as_bytes()).sign(&key).unwrap();

    let pubkey1_hash = key.publickey().as_hash();
    let pubkey2 = ECDSAPublicKey::from_hash(&pubkey1_hash).unwrap();
    Hasher::new_with_message(msg.as_bytes())
        .verify_signature(&sig, &pubkey2)
        .unwrap();
}

#[test]
fn test_hasher() {
    let mut h = Hasher::new();
    h.update_str("hello");
    let digest = h.get_hash_clone().digest().to_vec();
    let hexdigest = h.get_hash().to_hex();
    assert_eq!(
        hex::encode(&digest),
        hexdigest,
        "hexdigest matches digest"
    );
}
